### PR TITLE
Add logging levels and improve error handling

### PR DESCRIPTION
**Pull request by timfelle**
This pull request includes several commits that add logging levels and improve error handling in the codebase. 

The commits introduce new function for logging, `log_header`, and update `neko_error` to output in stderr. Additionally, a new environment variable `NEKO_LOG_FILE` is introduced to specify the log file name, with `stdout` used as the default if not set. The pull request also includes documentation updates for the new variable.

Additionally the json utilities have been extended to handle extraction of entire sub structures.

### CODE REVIEW

File: src/common/log.f90 [hunk 1]
```diff
@@ -149,19 +176,39 @@ subroutine log_message(this, msg, lvl)
...
+
   !> Write an error message to a log
   subroutine log_error(this, msg)
     class(log_t), intent(in) :: this
     character(len=*), intent(in) :: msg
 
     if (pe_rank .eq. 0) then
        call this%indent()
-       write(*, '(A,A,A)') '*** ERROR: ', trim(msg),'  ***'
+       write(this%unit_, '(A,A,A)') '*** ERROR: ', trim(msg),'  ***'
```
Comment (MartinKarp) [inline src/common/log.f90:211]: This is an error but does not kill the simulation then?

**Review by MartinKarp — COMMENTED** [review]
Looks good I think. If I understand correctly everything is the same, only we have expanded the functionality, but the defaults are as before?

File: src/common/log.f90 [hunk 2]
```diff
@@ -34,6 +34,7 @@
 module logger
   use comm, only : pe_rank
   use num_types, only : rp
+  use, intrinsic :: iso_fortran_env, only: stdout => output_unit
```
Comment (njansson) [inline src/common/log.f90:37]: Should we also use `error_unit` to be consistent

**Comment by timofeymukha** [comment]
Is there any difference between using an env variable and just redirecting to a file?

**Comment by timfelle** [comment]
> Is there any difference between using an env variable and just redirecting to a file?

Not all print statements are done through the logger. This way you can pipe official logs to a file and add your own prints to the terminal during debugging for example.

File: src/common/log.f90 (hunk 2 above)
Comment (timfelle) [reply src/common/log.f90:37]: Sure.

File: src/common/log.f90 (hunk 1 above)
Comment (timfelle) [reply src/common/log.f90:211]: Nope, this just prints to the log. I don't know that anyone is using it though.

Comment (njansson) [reply src/common/log.f90]: Not used, it has been like this for a while. But let's keep it if we ever unify stuff

File: src/common/log.f90 [hunk 3]
```diff
@@ -232,15 +305,12 @@ subroutine log_status(this, t, T_end)
...
-    call this%message('----------------------------------------------------------------', &
-                      NEKO_LOG_QUIET)
-    write(log_buf, '(A,E15.7,A,F6.2,A)') &
-    't = ', t, '                                  [ ',t_prog,'% ]'
-
+    call this%message(repeat('-', 64), NEKO_LOG_QUIET)
     call this%message(log_buf, NEKO_LOG_QUIET)
-    call this%message('----------------------------------------------------------------', &
-                      NEKO_LOG_QUIET)
+    call this%message(repeat('-', 64), NEKO_LOG_QUIET)
```
Comment (njansson) [inline src/common/log.f90:312]: This changes everything. `repeat` quality of life++ 🍻

**Review by njansson — APPROVED** [review]

File: src/common/log.f90 (hunk 3 above)
Comment (timfelle) [reply src/common/log.f90:312]: Oh yes, it is a lot better being able to use repeat. I guess we could avoid the loops in the indentation as well then 🤔

**Review by MartinKarp — APPROVED** [review]

**Pull request merged**